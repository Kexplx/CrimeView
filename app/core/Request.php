<?php

class Request
{
    public static function url()
    {
        return preg_replace('/\?.*$/', '', trim($_SERVER["REQUEST_URI"], '/'));
    }
}
