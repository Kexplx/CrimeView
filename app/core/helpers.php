<?php

function view(string $view, array $data = null)
{
    return require __DIR__ . "/../views/{$view}.view.php";
}
