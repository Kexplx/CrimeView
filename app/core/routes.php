<?php

return
    [
        "" => "PagesController@home",
        "home" => "PagesController@home",
        "routeData" => "RouteDataController@getRouteData",
        "compact" => "PagesController@compact",
    ];
