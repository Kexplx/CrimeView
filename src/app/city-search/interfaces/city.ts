export interface Position {
  lat: number;
  lng: number;
}

export interface City {
  placeId: string;
  name: string;
  position: Position;
}
