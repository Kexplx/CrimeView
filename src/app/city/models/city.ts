export interface City {
  placeId: string;
  name: string;
  lat: number;
  lng: number;
}
