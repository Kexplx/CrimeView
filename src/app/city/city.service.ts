import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface CityPrediction {
  placeId: string;
  name: string;
}

export interface City {
  placeId: string;
  name: string;
  state: string;
  state_short: string;
  lat: number;
  lng: number;
}

@Injectable()
export class CityService {
  constructor(
    private predictionsService: google.maps.places.AutocompleteService,
    private geoCoder: google.maps.Geocoder,
  ) {}

  private predictionsOptions = {
    componentRestrictions: { country: 'de' },
    types: ['(cities)'],
  };

  getCityPredictions(input: string): Observable<CityPrediction[]> {
    return new Observable(sub => {
      this.predictionsService.getPlacePredictions(
        { ...this.predictionsOptions, input },
        predictions => {
          const cityPredictions = predictions.map<CityPrediction>(p => ({
            name: p.description,
            placeId: p.place_id,
          }));

          sub.next(cityPredictions);
        },
      );
    });
  }

  getCity(placeId: string): Observable<City> {
    return new Observable(sub => {
      this.geoCoder.geocode({ placeId }, result => {
        const { address_components, geometry } = result[0];

        const city: City = {
          placeId,
          lat: geometry.location.lat(),
          lng: geometry.location.lng(),
          name: address_components[0].long_name,
          state: address_components[3].long_name,
          state_short: address_components[3].short_name,
        };

        sub.next(city);
      });
    });
  }
}
