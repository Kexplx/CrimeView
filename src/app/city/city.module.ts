import { NgModule } from '@angular/core';
import { CityListComponent } from './city-list/city-list.component';
import { SharedModule } from '../shared/shared.module';
import { CityService } from './city.service';
import { CityInputComponent } from './city-input/city-input.component';

@NgModule({
  declarations: [CityInputComponent, CityListComponent],
  imports: [SharedModule],
  providers: [
    CityService,
    google.maps.Geocoder,
    google.maps.places.AutocompleteService,
  ],
  exports: [CityListComponent],
})
export class CityModule {}
