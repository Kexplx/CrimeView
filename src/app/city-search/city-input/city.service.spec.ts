import { NgZone } from '@angular/core';
import { CityService } from './city.service';

type AutocompleteService = google.maps.places.AutocompleteService;
type Geocoder = google.maps.Geocoder;

const ngZoneStub = {
  run: (cb: () => void) => cb(),
};

const predictionServiceSpy = {
  getPlacePredictions: jest.fn((_: any, cb: any) =>
    cb([{ description: 't_desc', place_id: 't_placeId' }]),
  ),
};

const geocoderSpy = {
  geocode: jest.fn((_: any, cb: any) =>
    cb([
      {
        address_components: [{ long_name: 'cityName' }],
        geometry: { location: { lat: () => 1, lng: () => 0 } },
      },
    ]),
  ),
};

const cityService = new CityService(
  (predictionServiceSpy as unknown) as AutocompleteService,
  (geocoderSpy as unknown) as Geocoder,
  (ngZoneStub as unknown) as NgZone,
);

describe('#getCityPredictions', () => {
  it('calls #getPlacePredictions once', done => {
    cityService.getCityPredictions('').subscribe(() => {
      expect(predictionServiceSpy.getPlacePredictions.mock.calls.length).toEqual(1);

      done();
    });
  });

  it('returns observable with mapped predictions', done => {
    cityService.getCityPredictions('').subscribe(predictions => {
      expect(predictions).toEqual([{ name: 't_desc', placeId: 't_placeId' }]);

      done();
    });
  });
});

describe('#getCity', () => {
  it('should return an observable of cities', done => {
    cityService.getCity('placeId').subscribe(city => {
      expect(city).toEqual({ placeId: 'placeId', name: 'cityName', position: { lat: 1, lng: 0 } });

      done();
    });
  });
});
