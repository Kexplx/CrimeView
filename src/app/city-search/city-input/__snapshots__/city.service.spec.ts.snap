// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getCity should return an observable of cities 1`] = `
Object {
  "name": "cityName",
  "placeId": "placeId",
  "position": Object {
    "lat": 1,
    "lng": 0,
  },
}
`;

exports[`#getCityPredictions returns observable with mapped predictions 1`] = `
Array [
  Object {
    "name": "t_desc",
    "placeId": "t_placeId",
  },
]
`;
